### PR TITLE
fix(cli): pass ignore option to chokidar watcher to prevent watching ignored files

### PR DESCRIPTION
**Pull request by sudip-kumar-prasad**
FFixes #9565 by passing the parsed babel ignore options to chokidar's ignored configuration, preventing system limit errors for wide ignore scopes like `node_modules`.
Currently, when `@babel/cli` is run with the `--watch` flag, `chokidar` is initialized without knowing which files Babel is configured to ignore. This causes it to unnecessarily track large directories (such as `node_modules`), potentially leading to `ENOSPC` file watcher limit errors. This PR ensures that `babelOptions.ignore` is properly propagated down to `chokidar`'s `ignored` array.

### CODE REVIEW

**Comment by sudip-kumar-prasad** [comment]
Hi 👋 just following up on this PR.

All checks are passing, and I’ve addressed the review feedback.  
Would appreciate a review whenever you have time.

Thanks!

File: packages/babel-cli/src/babel/watcher.ts [hunk 1]
```diff
@@ -10,11 +12,48 @@ let watcher: FSWatcher;
...
+  watchQueue.clear();
+  hasStarted = false;
+
   isWatchMode = true;
 
   const { FSWatcher } = requireChokidar();
 
+  const ignoreArray = Array.isArray(ignore) ? ignore : ignore ? [ignore] : [];
+
+  const ignored = ignoreArray.length
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/watcher.ts]: This essentially duplicates the ignore handling logic in Babel-core and may introduce discrepancies between cli and core.

**Review by JLHwung — COMMENTED** [review]
Prior to 7.0, Babel-core uses `micromatch` to handle the `ignore` option. With https://github.com/babel/babel/pull/8327, we migrated to a much simpler match patterns.

The chokida 3.4 depended by Babel-cli also implements its own match item: https://github.com/paulmillr/chokidar/blob/c6d772a92f30004fdc437d00ecd54fbed041f30e/src/index.ts#L53. 

Could you analyzes any discrepancies between these two and ensure that passing through the `ignore` option must not lead to uncompiled files? It is acceptable if a file ignored by Babel-core is accepted by chokidar, but it is not acceptable if a compilable file is ignored by chokidar due to different `ignore` interpretation.

Out of curiosity, did you use LLM to assist opening this PR?

**Comment by sudip-kumar-prasad** [comment]
Thanks for the detailed feedback!

I’ve updated the implementation to rely on `loadPartialConfigSync`, so the watcher now uses the same ignore logic as Babel core instead of duplicating it.

This ensures that chokidar only ignores files when Babel itself would ignore them (i.e., when it returns `null`), so there’s no risk of skipping compilable files. It also correctly respects `.babelignore` and config-based ignore rules.

I’ve also added a small safety fallback so config errors don’t break the watcher.

Regarding your question — yes, I did use AI to assist while working on this, but I verified the changes manually.

Let me know if you’d like me to adjust anything further

**Comment by sudip-kumar-prasad** [comment]
Hi, just a quick follow-up on this.
I’ve addressed the previous feedback and ensured the watcher now uses Babel’s core config via loadPartialConfigSync.

Please let me know if anything else needs adjustment. Thanks!

File: packages/babel-cli/src/babel/watcher.ts [hunk 2]
```diff
@@ -10,11 +13,38 @@ let watcher: FSWatcher;
...
+  watchQueue.clear();
+  hasStarted = false;
+
   isWatchMode = true;
 
   const { FSWatcher } = requireChokidar();
 
+  const ignored = (watchedPath: string): boolean => {
+    try {
+      const config = loadPartialConfigSync({
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/watcher.ts:42]: Can we avoid invoking `loadPartialConfigSync` every time when chokidar determines if this file should be ignored?

**Comment by sudip-kumar-prasad** [comment]
I've optimized the watcher by implementing a result cache for the ignored callback. This ensuring loadPartialConfigSync is called only once per unique path during the initial scan/watch session, significantly reducing the overhead on large projects while maintaining full parity with Babel's ignore logic.

**Comment by sudip-kumar-prasad** [comment]
Thanks again for the detailed review!

I took a deeper look into the potential differences between Babel’s ignore handling and chokidar’s pattern matching.

To avoid any discrepancies, the watcher now relies on `loadPartialConfigSync` to determine whether a file should be ignored. This ensures we only ignore files when Babel itself would skip them (i.e., when it returns `null`), rather than depending on chokidar’s pattern matching.

Additionally:

* Implemented a cache for the ignored callback to avoid repeated config resolution for the same paths
* Ensured that even if chokidar’s pattern behavior differs, it won’t exclude files that Babel would compile

This way, we maintain correctness while still reducing unnecessary file watching.

Let me know if you’d like me to add tests specifically covering edge cases around ignore behavior.

File: packages/babel-cli/src/babel/watcher.ts [hunk 3]
```diff
@@ -10,11 +13,38 @@ let watcher: FSWatcher;
...
+        ...babelOptions,
+        filename: path.resolve(watchedPath),
+        showIgnoredFiles: false,
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/watcher.ts]: The option should be set to `true` and then we can check the `fileHandling` property. See https://github.com/babel/babel/pull/11907.

Can you add two new tests where the first one features a `.babelignore` file and the other one features the `ignore` option in a `.babelrc` file?

Comment (sudip-kumar-prasad) [reply packages/babel-cli/src/babel/watcher.ts]: Thanks for the review! I've updated the watcher to use showIgnoredFiles: true and check fileHandling as suggested.

I also added the two new tests for .babelignore and config-based ignore. While testing, I noticed that programmatic ignore patterns (like CLI --ignore) weren't correctly returning fileHandling: "ignored" in @babel/core, so I've included a fix for that as well.

All tests are passing now.

File: packages/babel-cli/test/fixtures/babel/dir-watch-ignore-config-file/options.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+  "args": ["src", "ignored", "--out-dir", ".", "--watch", "--config-file", "./babel-config.json"],
```
Comment (JLHwung) [inline packages/babel-cli/test/fixtures/babel/dir-watch-ignore-config-file/options.json:2]: This test is not about `.babelrc`, but rather a customized path `babel.config` file. See https://babel.dev/docs/config-files#file-relative-configuration

Comment (sudip-kumar-prasad) [reply packages/babel-cli/test/fixtures/babel/dir-watch-ignore-config-file/options.json:2]: Ah, you're totally right! I misunderstood how --config-file handles resolution.

I've just pushed a fix:

Renamed the test fixture to dir-watch-ignore-config-file.
Renamed the config file from .babelrc to babel-config.json.
Thanks for catching that and pointing me to the docs!
